// discord-trivia/dist/src/Classes/RootComponent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const discord_js_1 = require("discord.js");
class RootComponent {
    constructor(root) {
        this.reply = {
            [discord_js_1.InteractionType.MessageComponent]: (options) => {
                return this.entity.reply(options);
            },
            [discord_js_1.InteractionType.ApplicationCommand]: (options) => {
                return this.entity.reply(options);
            },
        };
        // The following error occures without ':any'
        // The inferred type of 'followUp' cannot be named without a reference to 'discord.js/node_modules/discord-api-types/v9'. This is likely not portable. A type annotation is necessary.
        this.followUp = {
            [discord_js_1.InteractionType.MessageComponent]: (options) => {
                return this.entity.reply(options);
            },
            [discord_js_1.InteractionType.ApplicationCommand]: (options) => {
                return this.entity.followUp(options);
            },
        };
        this.entity = root;
        if (root.content) {
            this.type = discord_js_1.InteractionType.MessageComponent;
        }
        else if (root.applicationId) {
            this.type = discord_js_1.InteractionType.ApplicationCommand;
        }
        else {
            throw new TypeError(`root argument must be of type Message or CommandInteraction`);
        }
        this.channel = root.channel;
        this.guild = root.guild;
        this.hostMember = root.member;
    }
}
exports.default = RootComponent;
