// discord-trivia/dist/src/Functions/prepareCustomQuestions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const open_trivia_db_1 = require("open-trivia-db");
/**
 * Takes an array of custom questions and parses them to be trivia game ready.
 * @param {CustomQuestion[]} questions - The custom questions.
 * @returns {Question[]} The parsed questions.
 */
function prepareCustomQuestions(questions) {
    return questions.map((q) => {
        const { value, difficulty, correctAnswer, incorrectAnswers, category } = q;
        const parsedQuestion = {
            value,
            category: new open_trivia_db_1.Category(category || "general knowledge").strictName,
            type: incorrectAnswers.length == 1 ? "boolean" : "multiple",
            difficulty: difficulty || "medium",
            correctAnswer,
            incorrectAnswers,
            allAnswers: open_trivia_db_1.OpenTDBUtil.shuffleArray([
                correctAnswer,
                ...incorrectAnswers,
            ]),
            checkAnswer: (arg) => {
                return arg.toLowerCase() == correctAnswer.toLowerCase();
            },
        };
        return parsedQuestion;
    });
}
exports.default = prepareCustomQuestions;
